pkg/services/screenshot/screenshot.go

```diff
@@ -0,0 +1,379 @@
+package screenshot
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"path"
+	"time"
+
+	gocache "github.com/patrickmn/go-cache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/grafana/grafana/pkg/components/imguploader"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/rendering"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const (
+	namespace = "grafana"
+	subsystem = "screenshot"
+)
+
+var (
+	DefaultTheme   = models.ThemeDark
+	DefaultTimeout = 15 * time.Second
+	DefaultHeight  = 500
+	DefaultWidth   = 1000
+)
+
+var (
+	ErrScreenshotsUnavailable = errors.New("screenshots unavailable")
+)
+
+// Screenshot represents a screenshot of a dashboard in Grafana.
+//
+// A screenshot can have a Path and an URL if the screenshot is stored on disk
+// and uploaded to a cloud storage service or made accessible via the Grafana
+// HTTP server.
+type Screenshot struct {
+	Path string
+	URL  string
+}
+
+// ScreenshotOptions are the options for taking a screenshot.
+type ScreenshotOptions struct {
+	DashboardUID string
+	PanelID      int64
+	Width        int
+	Height       int
+	Theme        models.Theme
+	Timeout      time.Duration
+}
+
+// SetDefaults sets default values for missing or invalid options.
+func (s ScreenshotOptions) SetDefaults() ScreenshotOptions {
+	if s.Width <= 0 {
+		s.Width = DefaultWidth
+	}
+	if s.Height <= 0 {
+		s.Height = DefaultHeight
+	}
+	switch s.Theme {
+	case models.ThemeDark, models.ThemeLight:
+	default:
+		s.Theme = DefaultTheme
+	}
+	if s.Timeout <= 0 {
+		s.Timeout = DefaultTimeout
+	}
+	return s
+}
+
+// ScreenshotService is an interface for taking screenshots.
+//go:generate mockgen -destination=mock.go -package=screenshot github.com/grafana/grafana/pkg/services/screenshot ScreenshotService
+type ScreenshotService interface {
+	Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error)
+}
+
+// BrowserScreenshotService takes screenshots using a headless browser.
+type BrowserScreenshotService struct {
+	ds                  dashboards.DashboardService
+	rs                  rendering.Service
+	screenshotDuration  prometheus.Histogram
+	screenshotFailures  prometheus.Counter
+	screenshotSuccesses prometheus.Counter
```

Comment:
:thinking: is it typical to inject monitoring right next to dependencies? (Not a golang expert, so I'm not quite sure what is canonical here)

~~Curious as to your thoughts on adding another decorator to wrap the metrics? `MonitoredScreenshotService` or something to that effect?~~

I just realized `ObservableScreenshotService` decorator exists with nearly the exact same metrics. Why duplicate the counters?

Reply:
We shouldn't - I'll dedupe

Reply:
In general it seems fine to me - I'm not big on decorators in Go. As JP mentioned in his review, it's fairly common to inject a capability into a general implementation that normally has "do nothing" versions of those capabilities:

```go
type Opts struct {
  authFunc func(req) authResul
  rateLimiter rate.Limiter
}